tsdb/head_test.go

```diff
@@ -5542,3 +5542,232 @@ func TestHeadCompactionWhileAppendAndCommitExemplar(t *testing.T) {
+}
+
+func TestOOOInsertDeleteQuery(t *testing.T) {
+	// Initialize a temporary directory and Write-Ahead Logs (WAL)
+	dir := t.TempDir()
+	wal, err := wlog.NewSize(nil, nil, filepath.Join(dir, "wal"), 32768, wlog.CompressionSnappy)
+	require.NoError(t, err)
+	oooWlog, err := wlog.NewSize(nil, nil, filepath.Join(dir, wlog.WblDirName), 32768, wlog.CompressionSnappy)
+	require.NoError(t, err)
+
+	// Set up Head options
+	opts := DefaultHeadOptions()
+	opts.ChunkDirRoot = dir
+	opts.OutOfOrderCapMax.Store(30)
+	opts.OutOfOrderTimeWindow.Store(120 * time.Minute.Milliseconds())
+
+	// Create a new Head
+	head, err := NewHead(nil, nil, wal, oooWlog, opts, nil)
+	require.NoError(t, err)
+	require.NoError(t, head.Init(0))
+
+	// Label the samples
+	lbls := labels.FromStrings("foo", "bar")
+
+	// Create a matcher to select samples by labels
+	matchers := []*labels.Matcher{
+		labels.MustNewMatcher(labels.MatchEqual, "foo", "bar"),
+	}
+
+	// Append OOO samples from t0 to t5
+	app := head.Appender(context.Background())
+	timestamps := []int64{10, 15, 20, 25, 30, 35}
+	for _, ts := range timestamps {
+		_, err = app.Append(0, lbls, ts, float64(ts))
+		require.NoError(t, err)
+	}
+	err = app.Commit()
+	require.NoError(t, err)
+
+	// Delete OOO samples t1 and t2
+	err = head.Delete(context.Background(), 15, 20, matchers...)
+	require.NoError(t, err)
+
+	// Try to insert OOO samples t1 and t2 again, this should fail or be ignored
+	app = head.Appender(context.Background())
+	app.Append(0, lbls, 15, 15.0)
+	app.Append(0, lbls, 20, 20.0)
+	err = app.Commit()
+	require.NoError(t, err)
+
+	// Query OOO samples from t0 to t5, expect t0, t3, t4, t5 (missing t1, t2 due to tombstoning)
+	q, err := NewBlockQuerier(head, head.MinTime(), head.MaxTime())
+	require.NoError(t, err)
+	defer q.Close()
+
+	set := q.Select(context.Background(), false, nil, matchers...)
+	valueType := set.Next()
+	require.NotEqual(t, chunkenc.ValNone, valueType)
+
+	series := set.At()
+	it := series.Iterator(nil)
+	remainT := []int64{}
+	for valueType := it.Next(); valueType != chunkenc.ValNone; valueType = it.Next() {
+		t, _ := it.At()
+		remainT = append(remainT, t)
+	}
+
+	expectedT := []int64{10, 25, 30, 35}
+	require.Equal(t, expectedT, remainT, "The queried timestamps are not as expected")
+
+	// Clean up resources
+	require.NoError(t, head.Close())
+}
+
+func TestOOOInsertDeleteCompactQuery(t *testing.T) {
+	dir := t.TempDir()
+	ctx := context.TODO()
+
+	opts := DefaultOptions()
+	opts.OutOfOrderCapMax = 30
+	opts.OutOfOrderTimeWindow = 300 * time.Minute.Milliseconds()
+
+	db, err := Open(dir, nil, nil, opts, nil)
+	require.NoError(t, err)
+	db.DisableCompactions()
+	defer func() {
+		require.NoError(t, db.Close())
+	}()
+
+	lbls := labels.FromStrings("foo", "bar")
+	matchers := []*labels.Matcher{
+		labels.MustNewMatcher(labels.MatchEqual, "foo", "bar"),
+	}
+
+	// Insert OOO samples from t0 to t5
+	app := db.Appender(ctx)
+	timestamps := []int64{10, 15, 20, 25, 30, 35}
```

Comment:
These are in-order samples rather than out of order samples - we need to add a sample with a newer timestamp before adding these samples.